### PR TITLE
Enhacement 9 ux for draggable panel

### PR DESCRIPTION
**Pull request by lionelB**
Au menu
- Handle responsive columns with using KeptColumns
- fix buggy drap 'n drop
- add a window config file for jest (not sure if jest config is also broke for mac / linux)

Hope it's KISS enought ;)

### CODE REVIEW

File: src/css/kept.css [hunk 1]
```diff
@@ -15,44 +15,29 @@ body {
...
-  }
-  @media (min-width: 1400px) {
-    .kept-list {
-      -webkit-columns: 5;
-      -moz-columns: 5;
-      columns: 5;
-    }
-  }
+
+.Columns {
```
Comment (n1k0) [inline src/css/kept.css]: please avoid uppercased class names

File: src/css/kept.css [hunk 2]
```diff
@@ -102,4 +87,4 @@ body {
     .todo-form li input[type="text"] {
       width: 145px;
     }
-  }
+  }
```
Comment (n1k0) [inline src/css/kept.css]: please ensure every file has a CR at the end

File: src/css/kept.css (hunk 1 above)
Comment (lionelB) [inline src/css/kept.css]: Sorry, kept-column seems better ?

File: src/css/kept.css (hunk 2 above)
Comment (lionelB) [inline src/css/kept.css]: I'll fix this !

File: src/js/components/KeptEntry.js [hunk 3]
```diff
@@ -60,7 +60,8 @@ var KeptEntry = React.createClass({
   handleOnDrop: function(event) {
     event.preventDefault();
     this.unhighlight();
-    this.props.move(event.dataTransfer.getData("text/plain"), this.props.key);
+    var newIndex = parseInt(event.dataTransfer.getData("text/plain"),10);
```
Comment (n1k0) [inline src/js/components/KeptEntry.js]: nit: space after comma

File: src/js/components/KeptItems.js [hunk 4]
```diff
@@ -3,10 +3,46 @@
...
 var DefaultContent = require("./DefaultContent");
 
+var times = function(nb) {
+  nb = parseInt(nb, 10) || 0;
+  var ar = new Array(nb);
+  while(nb-- > -1){
+    ar[nb] = nb;
+  }
+  return ar.slice();
+};
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: let's call this `range` and use something like this instead:

``` js
function range(n) {
  return Array.apply(0, Array(n)).map(function (_, i) { return i; });
}
```

This could be moved to a utils module too (and unit tested).

File: src/js/components/KeptItems.js [hunk 5]
```diff
@@ -3,10 +3,46 @@
...
+
 var KeepItems = React.createClass({
+  getInitialState: function() {
+    return {
+      columns: 1,
+      columnsWidth: 300
+    };
+  },
+  componentDidMount: function() {
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: nit: new line between methods

File: src/js/components/KeptItems.js [hunk 6]
```diff
@@ -3,10 +3,46 @@
...
+    // checking if we are inside a browser
+    if (!window) {
+      return;
+    }
+
+    window.addEventListener('resize', this.onResize);
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: let's extract this to a mixin please

File: src/js/components/KeptItems.js [hunk 7]
```diff
@@ -3,10 +3,46 @@
...
+  componentWillUnmount: function() {
+    // checking if we are inside a browser
+    if (!window) {
+      return;
+    }
+
+    window.removeEventListener('resize', this.onResize);
+  },
+  onResize: function(event){
+    var col = Math.floor(event.target.innerWidth/this.state.columnsWidth);
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: nit: space around operands please

Also, please test that `this.state.columnsWidth` doesn't equals zero

File: src/js/components/KeptItems.js [hunk 8]
```diff
@@ -3,10 +3,46 @@
...
+    this.setState({columns:col});
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: nit: space after colon

File: src/js/components/KeptItems.js [hunk 9]
```diff
@@ -14,16 +50,27 @@ var KeepItems = React.createClass({
...
-                    edit={this.props.edit}
-                    remove={this.props.remove}
-                    update={this.props.update}
-                    move={this.props.move} />;
+      <div className="kept-list">
+      {
+        times(this.state.columns).map( function(colNumber) {
+          var colItems = this.props.items.filter(function(item, i){
+            return i%this.state.columns === colNumber;
+          }.bind(this));
```
Comment (n1k0) [inline src/js/components/KeptItems.js]: no need to call bind(), just pass `this` as the second arg to `filter()`.

File: src/js/components/__tests__/KeptItems-test.js [hunk 10]
```diff
@@ -28,12 +28,15 @@ describe("KeptItems", function() {
...
+        {id: 5, type: "text", text: "text id #3"},
       ];
 
       var comp = TestUtils.renderIntoDocument(<KeptItems items={items} />);
 
-      var entries = TestUtils.scryRenderedComponentsWithType(comp, KeptEntry);
-      expect(entries.length).toEqual(items.length);
+      var entries = TestUtils.scryRenderedComponentsWithType(comp, KeptColumns);
+      //window.innerwidth  return 1024 so Math.floor(1024/300) => 4
+      expect(entries.length).toEqual(4);
```
Comment (n1k0) [inline src/js/components/__tests__/KeptItems-test.js]: This test is failing on the build.

File: src/css/kept.css [hunk 11]
```diff
@@ -15,44 +15,29 @@ body {
   margin-left: .35em;
 }
 
-.kept-list {
+.kept-list{
```
Comment (n1k0) [inline src/css/kept.css]: nit: space before curly bracket

File: package.json [hunk 12]
```diff
@@ -8,7 +8,8 @@
     "dist": "gulp dist",
     "lint": "jsxhint -c .jshintrc src/js",
     "start": "gulp",
-    "test": "jest --config jest-config.json && npm run lint"
+    "test": "jest --config jest-config.json && npm run lint",
+    "test-win": "jest --config jest-config-win.json && npm run lint"
```
Comment (n1k0) [inline package.json]: Out of curiosity, why is this needed?

File: src/js/components/KeptApp.js [hunk 13]
```diff
@@ -106,7 +106,11 @@ var KeptApp = React.createClass({
 
   move: function(fromIndex, toIndex) {
     var items = this.state.items.slice(0);
-    items.splice(toIndex, 0, items.splice(fromIndex, 1)[0]);
+    // switch items
+    var item = this.state.items[fromIndex];
+    items[fromIndex]  = items[toIndex];
+    items[toIndex]  = item;
```
Comment (n1k0) [inline src/js/components/KeptApp.js]: This feels like it could be extracted as a reusable helper (and unit tested).

File: package.json (hunk 12 above)
Comment (lionelB) [inline package.json]: cf https://github.com/facebook/jest/issues/100 
so i'll try to fix it the cleaner I can by adding anoter config file, hoping the issue will be close soon.

File: src/css/kept.css (hunk 1 above)
Comment (n1k0) [inline src/css/kept.css]: Yup

File: src/js/components/KeptApp.js (hunk 13 above)
Comment (lionelB) [inline src/js/components/KeptApp.js]: I'll add it with range (also array related)

File: src/js/components/KeptItems.js (hunk 9 above)
Comment (lionelB) [inline src/js/components/KeptItems.js]: does  space around operand rule apply to  `i%this.state.columns` ?

Comment (lionelB) [inline src/js/components/KeptItems.js]: also `this` arg also works for the `map()`

Comment (n1k0) [inline src/js/components/KeptItems.js]: > does space around operand rule apply to i%this.state.columns ?

yup, please

**Comment by lionelB** [comment]
J'ai tout fait mais le test du mixin de resize marche pas chez moi, jest affiche `---` dans le terminal et n’exécute pas le dernier test :(

j'ai mis à jour la branche sur mon repo https://github.com/lionelB/kept/tree/enhacement-9-UX-for-draggable-panel 
Si t'as un idée... parce jest, j'ai un peu du mal apparemment

**Comment by lionelB** [comment]
It was too late last time... take 3min to fix the test :D 
super green

**Comment by n1k0** [comment]
Merged (63ccb93b9f492e7b43c155c8f0200b9d65f44956)
